server/src/apub/fetcher.rs

```diff
@@ -43,36 +37,50 @@ use chrono::NaiveDateTime;
 static ACTOR_REFETCH_INTERVAL_SECONDS: i64 = 24 * 60 * 60;
 
 // Fetch nodeinfo metadata from a remote instance.
-fn _fetch_node_info(domain: &str) -> Result<NodeInfo, Error> {
+async fn _fetch_node_info(client: &Client, domain: &str) -> Result<NodeInfo, LemmyError> {
   let well_known_uri = Url::parse(&format!(
     "{}://{}/.well-known/nodeinfo",
     get_apub_protocol_string(),
     domain
   ))?;
-  let well_known = fetch_remote_object::<NodeInfoWellKnown>(&well_known_uri)?;
-  Ok(fetch_remote_object::<NodeInfo>(&well_known.links.href)?)
+
+  let well_known = fetch_remote_object::<NodeInfoWellKnown>(client, &well_known_uri).await?;
+  let nodeinfo = fetch_remote_object::<NodeInfo>(client, &well_known.links.href).await?;
+
+  Ok(nodeinfo)
 }
 
 /// Fetch any type of ActivityPub object, handling things like HTTP headers, deserialisation,
 /// timeouts etc.
-pub fn fetch_remote_object<Response>(url: &Url) -> Result<Response, Error>
+pub async fn fetch_remote_object<Response>(
+  client: &Client,
+  url: &Url,
+) -> Result<Response, LemmyError>
 where
   Response: for<'de> Deserialize<'de>,
 {
   if !is_apub_id_valid(&url) {
-    return Err(format_err!("Activitypub uri invalid or blocked: {}", url));
+    return Err(format_err!("Activitypub uri invalid or blocked: {}", url).into());
   }
-  // TODO: this function should return a future
+
   let timeout = Duration::from_secs(60);
-  let text = Request::get(url.as_str())
-    .header("Accept", APUB_JSON_CONTENT_TYPE)
-    .connect_timeout(timeout)
-    .timeout(timeout)
-    .body(())?
-    .send()?
-    .text()?;
-  let res: Response = serde_json::from_str(&text)?;
-  Ok(res)
+
+  let json = retry(|| {
+    client
```

Comment:
Fetching ActivityPub objects should probably be signed requests (authorized fetch), but I didn't know how to get the actor in here to make the request on their behalf.

Reply:
I agree, I'll make an issue for this.